ERR_IPC_CHANNEL_CLOSED error is logged when executing tests in @theia/core

In Appveyor, when compiling issue #610, The following error occurs:
Compilation fails for the build #610 in @theia/core
link: https://ci.appveyor.com/project/kittaakos/theia/build/1.0.954
 ```
@theia/core: The express server worker 1 failed: { Error [ERR_IPC_CHANNEL_CLOSED]: channel closed\n    at ChildProcess.target.send (internal/child_process.js:590:16)\n    at Worker.send (internal/cluster/worker.js:54:28)\n    at WorkerMessageWriter.send (C:\projects\theia\packages\core\src\node\cluster\worker-writer.ts:6:6143)\n    at WorkerMessageWriter.AbstractStreamMessageWriter.write (C:\projects\theia\node_modules\vscode-ws-jsonrpc\src\stream\writer.ts:13:14)\n    at reply (C:\projects\theia\node_modules\vscode-jsonrpc\lib\main.js:301:27)\n    at C:\projects\theia\node_modules\vscode-jsonrpc\lib\main.js:363:25\n    at <anonymous> [Symbol(code)]: 'ERR_IPC_CHANNEL_CLOSED' }
```



<b>Update:</b>
It does not cause a build failure, but the error exists on all CI platforms.

Although that is a test failure, I cannot see anything related to `ERR_IPC_CHANNEL_CLOSED`. Do I miss something?

```
@theia/preferences: Killed
@theia/preferences: npm ERR! code ELIFECYCLE
@theia/preferences: npm ERR! errno 137
@theia/preferences: npm ERR! @theia/preferences@0.2.0 test: `theiaext test`
@theia/preferences: npm ERR! Exit status 137
@theia/preferences: npm ERR! 
@theia/preferences: npm ERR! Failed at the @theia/preferences@0.2.0 test script.
@theia/preferences: npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
@theia/preferences: npm ERR! A complete log of this run can be found in:
@theia/preferences: npm ERR!     /home/travis/.npm/_logs/2017-10-25T20_03_24_746Z-debug.log
lerna ERR! execute Error: Command failed: npm run test
lerna ERR! execute Killed
lerna ERR! execute npm ERR! code ELIFECYCLE
lerna ERR! execute npm ERR! errno 137
lerna ERR! execute npm ERR! @theia/preferences@0.2.0 test: `theiaext test`
lerna ERR! execute npm ERR! Exit status 137
lerna ERR! execute npm ERR! 
lerna ERR! execute npm ERR! Failed at the @theia/preferences@0.2.0 test script.
lerna ERR! execute npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
lerna ERR! execute 
lerna ERR! execute npm ERR! A complete log of this run can be found in:
lerna ERR! execute npm ERR!     /home/travis/.npm/_logs/2017-10-25T20_03_24_746Z-debug.log
lerna ERR! execute 
lerna ERR! execute 
lerna ERR! execute > @theia/preferences@0.2.0 test /home/travis/build/theia-ide/theia/packages/preferences
lerna ERR! execute > theiaext test
lerna ERR! execute 
lerna ERR! execute 
lerna ERR! execute     at Promise.all.then.arr (/home/travis/build/theia-ide/theia/node_modules/execa/index.js:236:11)
lerna ERR! execute     at <anonymous>
lerna ERR! execute     at process._tickCallback (internal/process/next_tick.js:188:7)
lerna ERR! execute  callback with error
lerna ERR! execute  { Error: Command failed: npm run test
lerna ERR! execute Killed
lerna ERR! execute npm ERR! code ELIFECYCLE
lerna ERR! execute npm ERR! errno 137
lerna ERR! execute npm ERR! @theia/preferences@0.2.0 test: `theiaext test`
lerna ERR! execute npm ERR! Exit status 137
lerna ERR! execute npm ERR! 
lerna ERR! execute npm ERR! Failed at the @theia/preferences@0.2.0 test script.
lerna ERR! execute npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
lerna ERR! execute 
lerna ERR! execute npm ERR! A complete log of this run can be found in:
lerna ERR! execute npm ERR!     /home/travis/.npm/_logs/2017-10-25T20_03_24_746Z-debug.log
lerna ERR! execute 
lerna ERR! execute 
lerna ERR! execute > @theia/preferences@0.2.0 test /home/travis/build/theia-ide/theia/packages/preferences
lerna ERR! execute > theiaext test
lerna ERR! execute 
lerna ERR! execute 
lerna ERR! execute     at Promise.all.then.arr (/home/travis/build/theia-ide/theia/node_modules/execa/index.js:236:11)
lerna ERR! execute     at <anonymous>
lerna ERR! execute     at process._tickCallback (internal/process/next_tick.js:188:7)
lerna ERR! execute   stack: 'Error: Command failed: npm run test\nKilled\nnpm ERR! code ELIFECYCLE\nnpm ERR! errno 137\nnpm ERR! @theia/preferences@0.2.0 test: `theiaext test`\nnpm ERR! Exit status 137\nnpm ERR! \nnpm ERR! Failed at the @theia/preferences@0.2.0 test script.\nnpm ERR! This is probably not a problem with npm. There is likely additional logging output above.\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /home/travis/.npm/_logs/2017-10-25T20_03_24_746Z-debug.log\n\n\n> @theia/preferences@0.2.0 test /home/travis/build/theia-ide/theia/packages/preferences\n> theiaext test\n\n\n    at Promise.all.then.arr (/home/travis/build/theia-ide/theia/node_modules/execa/index.js:236:11)\n    at <anonymous>\n    at process._tickCallback (internal/process/next_tick.js:188:7)',
lerna ERR! execute   code: 137,
lerna ERR! execute   killed: false,
lerna ERR! execute   stdout: '\n> @theia/preferences@0.2.0 test /home/travis/build/theia-ide/theia/packages/preferences\n> theiaext test\n\n',
lerna ERR! execute   stderr: 'Killed\nnpm ERR! code ELIFECYCLE\nnpm ERR! errno 137\nnpm ERR! @theia/preferences@0.2.0 test: `theiaext test`\nnpm ERR! Exit status 137\nnpm ERR! \nnpm ERR! Failed at the @theia/preferences@0.2.0 test script.\nnpm ERR! This is probably not a problem with npm. There is likely additional logging output above.\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /home/travis/.npm/_logs/2017-10-25T20_03_24_746Z-debug.log\n',
lerna ERR! execute   failed: true,
lerna ERR! execute   signal: null,
lerna ERR! execute   cmd: 'npm run test',
lerna ERR! execute   timedOut: false }
lerna WARN complete Waiting for 18 child processes to exit. CTRL-C to exit immediately.
@theia/python:   python package
```

And later:

```
@theia/extension-manager: Building the app... 
@theia/core:     1) restart
@theia/core: The express server worker 1 has been disconnected.
@theia/core: The express server worker 1 has been stopped.
@theia/core:   server-process
@theia/core: The express server worker 3 is ready to accept messages.
@theia/core: The express server worker has been restarted.
@theia/core: The express server worker 1 failed: Error [ERR_IPC_CHANNEL_CLOSED]: channel closed\n    at ChildProcess.target.send (internal/child_process.js:590:16)\n    at Worker.send (internal/cluster/worker.js:54:28)\n    at WorkerMessageWriter.send (/home/travis/build/theia-ide/theia/packages/core/src/node/cluster/worker-writer.ts:6:6161)\n    at WorkerMessageWriter.AbstractStreamMessageWriter.write (/home/travis/build/theia-ide/theia/node_modules/vscode-ws-jsonrpc/src/stream/writer.ts:13:14)\n    at reply (/home/travis/build/theia-ide/theia/node_modules/vscode-jsonrpc/lib/main.js:301:27)\n    at /home/travis/build/theia-ide/theia/node_modules/vscode-jsonrpc/lib/main.js:363:25\n    at <anonymous>
@theia/core: The express server worker 3 has been disconnected.
@theia/core: The express server worker 3 has been stopped.
@theia/core:     ✓ start (3679ms)
```

Did you mean this at the second attempt?
```
@theia/core: The express server worker 2 has been started.
@theia/git:       ✓ modified in index and in working directory (279ms)
@theia/git:   9 passing (8s)
@theia/core:     1) restart
@theia/extension-manager: Added: /home/travis/build/theia-ide/theia/packages/extension-manager/application-project-test-temp/117925-12193-3fug5u.34hd8/package.json
@theia/core: The express server worker 1 has been disconnected.
@theia/core: The express server worker 1 has been stopped.
@theia/core:   0 passing (16ms)
@theia/core: The express server worker 2 has been disconnected.
@theia/core: The express server worker failed to restart.
@theia/core: The express server worker 2 has been stopped.
@theia/core: The express server worker 1 failed: Error [ERR_IPC_CHANNEL_CLOSED]: channel closed\n    at ChildProcess.target.send (internal/child_process.js:590:16)\n    at Worker.send (internal/cluster/worker.js:54:28)\n    at WorkerMessageWriter.send (/home/travis/build/theia-ide/theia/packages/core/src/node/cluster/worker-writer.ts:6:6161)\n    at WorkerMessageWriter.AbstractStreamMessageWriter.write (/home/travis/build/theia-ide/theia/node_modules/vscode-ws-jsonrpc/src/stream/writer.ts:13:14)\n    at replyError (/home/travis/build/theia-ide/theia/node_modules/vscode-jsonrpc/lib/main.js:310:27)\n    at /home/travis/build/theia-ide/theia/node_modules/vscode-jsonrpc/lib/main.js:370:29\n    at <anonymous>\n    at process._tickCallback (internal/process/next_tick.js:188:7)
@theia/preferences:   compound-preference-server
@theia/preferences:     ✓ register a client
@theia/preferences:   json-preference-server
@theia/preferences:     Json pref server client
@theia/preferences:       ✓ Register a client and change the value
@theia/preferences:   2 passing (49ms)
@theia/core:     1) start
@theia/core:   file-uri
@theia/core:     ✓ create -> fsPath -> create should be symmetric
@theia/core:     ✓ fsPath -> create -> fsPath should be symmetric
@theia/core:     ✓ from /
@theia/core:     ✓ from //
@theia/core:     ✓ from c:
@theia/core:     ✓ from /c:
@theia/core:     ✓ from /c:/
@theia/core:   55 passing (11s)
@theia/core:   1 failing
@theia/core:   1) master-process start:
@theia/core:      Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
@theia/core:   
@theia/core: npm ERR! code ELIFECYCLE
@theia/core: npm ERR! errno 1
@theia/core: npm ERR! @theia/core@0.2.0 test: `theiaext test`
@theia/core: npm ERR! Exit status 1
@theia/core: npm ERR! 
@theia/core: npm ERR! Failed at the @theia/core@0.2.0 test script.
@theia/core: npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
@theia/core: npm ERR! A complete log of this run can be found in:
@theia/core: npm ERR!     /home/travis/.npm/_logs/2017-10-25T20_05_05_062Z-debug.log
lerna ERR! execute Error: Command failed: npm run test
lerna ERR! execute The express server worker failed to restart.
lerna ERR! execute The express server worker 1 failed: Error [ERR_IPC_CHANNEL_CLOSED]: channel closed
lerna ERR! execute     at ChildProcess.target.send (internal/child_process.js:590:16)
lerna ERR! execute     at Worker.send (internal/cluster/worker.js:54:28)
lerna ERR! execute     at WorkerMessageWriter.send (/home/travis/build/theia-ide/theia/packages/core/src/node/cluster/worker-writer.ts:6:6161)
lerna ERR! execute     at WorkerMessageWriter.AbstractStreamMessageWriter.write (/home/travis/build/theia-ide/theia/node_modules/vscode-ws-jsonrpc/src/stream/writer.ts:13:14)
lerna ERR! execute     at replyError (/home/travis/build/theia-ide/theia/node_modules/vscode-jsonrpc/lib/main.js:310:27)
lerna ERR! execute     at /home/travis/build/theia-ide/theia/node_modules/vscode-jsonrpc/lib/main.js:370:29
lerna ERR! execute     at <anonymous>
lerna ERR! execute     at process._tickCallback (internal/process/next_tick.js:188:7)
lerna ERR! execute npm ERR! code ELIFECYCLE
lerna ERR! execute npm ERR! errno 1
lerna ERR! execute npm ERR! @theia/core@0.2.0 test: `theiaext test`
lerna ERR! execute npm ERR! Exit status 1
lerna ERR! execute npm ERR! 
lerna ERR! execute npm ERR! Failed at the @theia/core@0.2.0 test script.
lerna ERR! execute npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
lerna ERR! execute 
lerna ERR! execute npm ERR! A complete log of this run can be found in:
lerna ERR! execute npm ERR!     /home/travis/.npm/_logs/2017-10-25T20_05_05_062Z-debug.log
lerna ERR! execute 
lerna ERR! execute 
```

--- a/packages/core/src/node/cluster.spec.ts
+++ b/packages/core/src/node/cluster.spec.ts
@@ -6,79 +6,60 @@
  */
 
 // tslint:disable:no-console
-import { assert, use } from 'chai';
-use(require('chai-as-promised'));
+import * as path from 'path';
 import * as cluster from 'cluster';
-import { clusterRemoteMasterProcessFactory, ServerProcess } from './cluster';
 import { MasterProcess } from './cluster/master-process';
 
 process.on('unhandledRejection', (reason, promise) => {
     throw reason;
 });
 
-if (cluster.isMaster) {
-    describe('master-process', () => {
+describe('master-process', () => {
 
-        function prepareWorkerTest(test: string) {
-            // https://github.com/Microsoft/vscode/issues/3201
-            const execArgv = process.execArgv.reduce((result, arg) => {
-                result.push(arg.replace('-brk', ''));
-                return result;
-            }, [] as string[]);
-            const testSettings: cluster.ClusterSettings = {
-                execArgv,
-                args: [...process.argv.slice(2), '--grep', test],
-                // https://github.com/Microsoft/vscode/issues/19750
-                stdio: ['ipc', 'pipe', 'pipe']
-            };
-            cluster.setupMaster(testSettings);
-        }
-
-        const forwardWorkerTestOutput = (worker: cluster.Worker) => {
-            worker.process.stdout.pipe(process.stdout);
-            worker.process.stderr.pipe(process.stderr);
+    function prepareTestWorker(job: string) {
+        const testSettings: cluster.ClusterSettings = {
+            exec: path.resolve(__dirname, '../../lib/node/test/cluster-test-worker.js'),
+            execArgv: [],
+            args: [job],
+            stdio: ['ipc', 1, 2]
         };
-        before(() => {
-            cluster.on('fork', forwardWorkerTestOutput);
-        });
-        after(() => {
-            cluster.removeListener('fork', forwardWorkerTestOutput);
-        });
+        cluster.setupMaster(testSettings);
+    }
 
-        let originalSettings: cluster.ClusterSettings;
-        beforeEach(() => originalSettings = cluster.settings);
-        afterEach(() => cluster.setupMaster(originalSettings));
+    let originalSettings: cluster.ClusterSettings;
+    beforeEach(() => originalSettings = cluster.settings);
+    afterEach(() => cluster.setupMaster(originalSettings));
 
-        it('start', async function () {
-            this.timeout(10000);
-            const master = new MasterProcess();
+    /**
+     * Tests restarting of workers by the master process:
+     * 1. the master process starts the server worker with `restart` job
+     * 2. Testing failed restart
+     *   2.1 the first worker sends `restart` request to the master
+     *   2.2 the master tries to start the second worker with `timeout next worker` job
+     *   2.3 the second worker fails because such job does not exist
+     *   2.4 the master throws the error to the first worker
+     *   2.5 the first worker checks that the error is received
+     * 3. Testing successful restart
+     *   3.1 the first worker sends `restart` request again to the master
+     *   3.2 the master tries to start the third worker with `restarted` job
+     *   3.3 the third worker is successfully initialized and then exits
+     *   3.4 the master worker returns to the first worker
+     *   3.5 the first worker exits
+     */
+    it('start', async function () {
+        this.timeout(10000);
+        const master = new MasterProcess();
 
-            prepareWorkerTest('restart');
-            const restartWorker = master.start();
+        prepareTestWorker('restart');
+        const restartWorker = master.start();
 
-            prepareWorkerTest('timeout next worker');
-            await master.restarting;
-            prepareWorkerTest('restarted');
+        prepareTestWorker('timeout next worker');
+        await master.restarting;
+        prepareTestWorker('restarted');
 
-            const restartedWorker = await master.restarted;
-            await assert.isFulfilled(restartWorker.exit);
-            await assert.isFulfilled(restartedWorker.exit);
-        });
-
+        const restartedWorker = await master.restarted;
+        await restartWorker.exit;
+        await restartedWorker.exit;
     });
-} else {
-    describe('server-process', () => {
 
-        it('restart', async function () {
-            const server = new ServerProcess(clusterRemoteMasterProcessFactory);
-            await assert.isRejected(server.restart());
-            await assert.isFulfilled(server.restart());
-        });
-
-        it('restarted', async function () {
-            const server = new ServerProcess(clusterRemoteMasterProcessFactory);
-            await assert.isFulfilled(server.kill());
-        });
-
-    });
-}
+});

--- a/packages/core/src/node/test/cluster-test-worker.ts
+++ b/packages/core/src/node/test/cluster-test-worker.ts
@@ -1,0 +1,46 @@
+/*
+ * Copyright (C) 2017 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+// tslint:disable:no-console
+import 'reflect-metadata';
+import { clusterRemoteMasterProcessFactory, ServerProcess } from '../cluster';
+
+const jobs: { [id: string]: (() => Promise<void>) | undefined } = {
+    restart: async () => {
+        const server = new ServerProcess(clusterRemoteMasterProcessFactory);
+        let firstRestartFailed = true;
+        try {
+            await server.restart();
+            firstRestartFailed = false;
+        } catch (e) {
+            if ((e as Error).message.indexOf('failed to restart') === -1) {
+                throw e;
+            }
+        }
+        if (firstRestartFailed) {
+            await server.restart();
+        } else {
+            throw new Error('first restart should fail');
+        }
+    },
+    restarted: async () => {
+        const server = new ServerProcess(clusterRemoteMasterProcessFactory);
+        await server.kill();
+    }
+};
+
+const id = process.argv[process.argv.length - 1];
+const job = jobs[id];
+if (job) {
+    job().then(() =>
+        process.exit(0),
+        reason => {
+            console.error(`Test worker: '${id}' failed`, reason);
+            process.exit(1);
+        }
+    );
+}

